crates/client/data-availability/src/lib.rs

```diff
@@ -0,0 +1,159 @@
+            let mut nonces: HashMap<&[u8], &[u8]> = HashMap::new();
+            let mut storage_diffs: HashMap<&[u8], StorageWrites> = HashMap::new();
+
+            // Locate and encode the storage change
+            for event in storage_event.changes.iter() {
+                let mut prefix = event.1.0.as_slice();
+                let mut key: &[u8] = &[];
+                if prefix.len() > 32 {
+                    let raw_split = prefix.split_at(32);
+                    prefix = raw_split.0;
+                    key = raw_split.1;
+                }
+
+                if prefix == *utils::SN_NONCE_PREFIX {
+                    if let Some(data) = event.2 {
+                        nonces.insert(key, data.0.as_slice());
+                    }
+                }
+
+                if prefix == *utils::SN_STORAGE_PREFIX {
+                    if let Some(data) = event.2 {
+                        // first 32 bytes = contract address, second 32 bytes = storage variable
+                        let write_split = key.split_at(32);
+
+                        storage_diffs
+                            .entry(write_split.0)
+                            .and_modify(|v| v.push((write_split.1, data.0.as_slice())))
+                            .or_insert(vec![(write_split.1, data.0.as_slice())]);
+                    }
+                }
+            }
+
+            let state_diff = utils::pre_0_11_0_state_diff(storage_diffs, nonces);
+
+            // Store the DA output from the SN OS
+            if let Err(db_err) = madara_backend.da().store_state_diff(&storage_event.block, state_diff) {
+                log::error!("db err: {db_err}");
+            };
+
+            match da_mode {
+                DaMode::Validity => {
+                    // Submit the StarkNet OS PIE
+                    if let Ok(job_resp) = sharp::submit_pie("test") {
+                        log::info!("Job Submitted: {}", job_resp.cairo_job_key);
+                        // Store the cairo job key
+                        let _res = madara_backend.da().update_cairo_job(&storage_event.block, job_resp.cairo_job_key);
+                    }
+                }
+                _ => {
+                    log::info!("don't prove in remaining DA modes")
+                }
+            }
+        }
+    }
+}
+
+impl<B, C> DataAvailabilityWorker<B, C>
+where
+    B: BlockT,
+    C: ProvideRuntimeApi<B>,
+    C: BlockchainEvents<B> + 'static,
+{
+    // pub async fn update_state(client: Arc<C>, madara_backend: Arc<mc_db::Backend<B>>) {
+    pub async fn update_state(
+        da_client: Box<dyn DaClient + Send + Sync>,
+        client: Arc<C>,
+        madara_backend: Arc<mc_db::Backend<B>>,
+    ) {
+        let mut notification_st = client.import_notification_stream();
+
+        while let Some(notification) = notification_st.next().await {
+            // Query last written state
+            let _last_published_state = match da_client.last_published_state().await {
```

Comment:
Why is the return from this query never used? Is the query still necessary?

Reply:
Or is it only needed for `Validity` mode?

Reply:
`// TODO: this value will be used to ensure the correct state diff is being written in Validity mode`

added as comment